### DISCUSSION TITLE
Multiple MemoryStream objects

### DISCUSSION THREAD

**Discussion by bobthechemist** [pschatzmann/arduino-audio-tools#572] [Q&A] [answered]
Is there an example of handling multiple MemoryStream objects? In my project, I would like to cycle through a series of clips when the user presses a button. I am looking at `streams-memory_mp3-analog` as an example; however it seems like I would need to declare and then update the three streams as well as the copier within the loop as opposed to how it is organized in the example. Before I dig myself into a hole, I wanted to check if there are better examples for my intended application.

**Comment by pschatzmann** [comment, marked as answer]
Two thoughts that might help
1) You have 1 copier and just just add the logic to define the copy source dependent on your requirements. (check for the parameters available for the begin method in StreamCopy)
2) You use the functionality of the AudioPlayer and implement your custom data source (subclass of AudioSource) or using the AudioSourceCallback that would return the "right" stream as requred by your logic 

¨

**Comment by bobthechemist** [comment]
Thanks for the quick response, Phil,

As I'm learning the system, I think the most appropriate approach is to store the small MP3 files on SPIFFS. Since I'm not using I2S, I've modified the example to sink to analog.

```
#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioSourceSPIFFS.h"

const char *startFilePath="/";
const char *ext="mp3";

AudioSourceSPIFFS source(startFilePath, ext);
AnalogAudioStream sink;
MP3DecoderHelix decoder;
AudioPlayer player (source, sink, decoder);

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // begin processing
  auto cfg = sink.defaultConfig();
  sink.begin(cfg);
  
  player.setMetadataCallback(printMetaData);
  player.begin();
}


void loop(){
  player.copy();
}
```
Three mp3 files have been uploaded to the filesystem and the audiotools info messages indicate they are being seen; however there is a timeout, which makes me think they are not read properly.  Here is what I believe to be the relevant info/warning messages:

```
[I] AudioSourceSPIFFS.h : 45 - selectStream: 0
[I] AudioSourceSPIFFS.h : 49 - Using file //mouse_01.mp3
[I] AudioCopy.h : 69 - buffer_size=1024
[W] AudioPlayer.h : 388 - -> timeout - moving by 1
[I] AudioSourceSPIFFS.h : 40 - nextStream: 1
[I] AudioSourceSPIFFS.h : 45 - selectStream: 1
[I] AudioSourceSPIFFS.h : 49 - Using file //mouse_02.mp3
```
Does the timeout suggest that the mp3 is not being read properly or has the incorrect format?

**Comment by pschatzmann** [comment]
What happens when you put the files into a subdirectory ?  The two slashes at the beginning might cause some issues...

**Comment by bobthechemist** [comment]
Moving the files to a subdirectory and then referencing that directory with `const char *startFilePath="/sounds";` seems to have done the trick. Thanks again for the quick turnaround.
